script/src/scheduler.rs

```diff
@@ -845,35 +873,33 @@ where
     }
 
     // Create a new VM instance with syscalls attached
-    fn create_dummy_vm(&self, id: &VmId) -> Result<(MachineContext<DL>, Machine), Error> {
+    fn create_dummy_vm(&self, id: &VmId) -> Result<(VmContext<DL>, Machine), Error> {
         // The code here looks slightly weird, since I don't want to copy over all syscall
         // impls here again. Ideally, this scheduler package should be merged with ckb-script,
         // or simply replace ckb-script. That way, the quirks here will be eliminated.
-        let version = self.script_version;
+        let version = &self.sg_data.script_version;
         let core_machine = CoreMachineType::new(
             version.vm_isa(),
             version.vm_version(),
             // We will update max_cycles for each machine when it gets a chance to run
             u64::MAX,
         );
-        let snapshot2_context = Arc::new(Mutex::new(Snapshot2Context::new(self.tx_data.clone())));
-        let mut syscalls_generator = self.syscalls_generator.clone();
-        syscalls_generator.vm_id = *id;
-        let mut machine_context = MachineContext::new(self.tx_data.clone());
-        machine_context.base_cycles = Arc::clone(&self.syscalls_generator.base_cycles);
-        machine_context.snapshot2_context = Arc::clone(&snapshot2_context);
+        let vm_data = Arc::new(VmData {
```

Comment:
I don't see the need for a VmData structure here, it's just wrapped struct which add a vm_id field compared to SgData, and this field is only used in generate_ckb_syscalls fn to pass vm_id to the relevant syscall constructor, e.g.. 
```
ExecV2::new(vm_data, vm_context)
ProcessID::new(vm_data)
```
We can add a vm_id parameter to generate_ckb_syscalls fn, then change the syscall constructor to
```
ExecV2::new(vm_id, vm_context)
ProcessID::new(vm_id)
```

This will simplify the code, e.g. reduce the use of Arc::new: ``diff

``diff
- let vm_data = Arc::new(VmData {
- sg_data: Arc::clone(&self.sg_data),
- vm_id: *id, .
- });
+ let sg_data = Arc::clone(&self.sg_data),
```

Reply:
Done